### ISSUE TITLE
config.txt customisation overwrites more variables than it should

### ISSUE THREAD

**Issue by hhromic** [WebPlatformForEmbedded/meta-wpe#177]
Hi guys,
I noticed a small bug in the config.txt customisation recipes:

* https://github.com/WebPlatformForEmbedded/meta-wpe/blob/master/recipes-bsp/bootfiles/rpi-config_git.bbappend
* https://github.com/WebPlatformForEmbedded/meta-wpe/blob/master/recipes-raspberrypi/bootfiles/rpi-config_%25.bbappend

This is also present in the `meta-raspberrypi` layer and I reported it there as well: https://github.com/agherzan/meta-raspberrypi/issues/181

In summary, the `sed` lines need a `=` sign after the variable names, so other variables that begin with the same name do not get customised accidentally. For example `gpu_freq` also replaces `gpu_freq_min` with `gpu_freq=500`, making the config.txt file containing this customisation twice.

    # these lines
    sed -i '/#gpu_freq/ c\gpu_freq=500' ${DEPLOYDIR}/bcm2835-bootfiles/config.txt

    # should be like these lines (note the = sign)
    sed -i '/#gpu_freq=/ c\gpu_freq=500' ${DEPLOYDIR}/bcm2835-bootfiles/config.txt

It's a small and easy fix =).

**Comment by wouterlucas** [comment]
Done! Thanks for reporting :)

**Comment by hhromic** [comment]
no problem!
I noticed you only fixed the recipe in `recipes-bsp/bootfiles/rpi-config_git.bbappend`.
I belive the recipe in `recipes-raspberrypi/bootfiles/rpi-config_%25.bbappend` should also be updated for this change, to be future-proof. For example if they ever add `hdmi_group_XYZ` or `hdmi_mode_XYZ` options.
Thanks for the great work!

**Comment by wouterlucas** [comment]
You're welcome!

That %25 file is not in the repo:
https://github.com/WebPlatformForEmbedded/meta-wpe/tree/master/recipes-bsp/bootfiles

Maybe that's a local file? I'm not sure if they're going to add versioning to rpi-config, else we could do a % to catch 'em. But for now _git seems to be the way.

**Comment by hhromic** [comment]
The file is not in the same directory than the recipe you fixed, but is under the `recipes-raspberrypi` folder, not in `recipes-bsp`. Here is the URL for it again for reference:

https://github.com/WebPlatformForEmbedded/meta-wpe/blob/master/recipes-raspberrypi/bootfiles/rpi-config_%25.bbappend

Looks like that recipe's only job is to set the `hdmi_*` configuration variables.

**Comment by wouterlucas** [comment]
Oh right forgot that's there, I think that's a left over from how meta-raspberrypi was structured initially/while ago and should be merged into recipes-bsp.

**Comment by hhromic** [comment]
That makes a lot of sense. I got thinking that indeed those looked "out of place".

I even noticed that the mainstream `meta-raspberrypi` layer now has a configuration for `disable_overscan` as well.
It also has an extra parameters facility that could render the entire recipe in `meta-wpe` obsolete. All the details are here:
http://meta-raspberrypi.readthedocs.io/en/latest/extra-build-config.html

For example, to add the `gpu_freq` and `boot_delay`parameters, we should do something like this:

    RPI_EXTRA_CONFIG = ' \n \
        # gpu frequency and boot delay \n \
        gpu_freq=500 \n \
        boot_delay=0 \n \
        '

This should be added in the recipe in the `do_deploy_prepend()` function.

I'm a bit busy this week but towards the weekend I can work out these simplifications/optimisations, test them and send a pull request to you. How does that sound?

**Comment by wouterlucas** [comment]
Yes that works, I tried that. Its just the newlines didn't work for me and it all ended up as a 1 line in config.txt for me.

Haven't tried if it actually gets picked up by an physical RPI. At that point my OCD kicked in, shivered at the \n characters in there and I went back to sed'ing the parameters :neckbeard:

**Comment by hhromic** [comment]
I will give it another try this weekend. My OCD also kicks with these things :).
As I'm very new to Yocto, I have a question: what is the best practice for quickly rebuilding the image after you modify recipes? i.e. just rebuild the changes.
If I modify recipes in other layers (e.g. `meta-raspberrypi`) do I have to clean the build completely, re-add all layers and rebuild? this takes ages :(
Thanks and happy to help to this projcet!

**Comment by wouterlucas** [comment]
No not at all, a clean build is only really required when you're switching Yocto versions (like going from master to morty). Bitbake will see which recipes you've changed and rebuild accordingly, sysroot is created out of the packages that are in the dependency chain and bitbake will try to ensure it is always "clean" (unlike systems like Buildroot).

Though like most systems its good to go back to a clean state once every while, I usually wipe my sstate cache  and tmp dir once a week or so.

The only downside is that Yocto has ccache disabled, which hurts if you need to rebuild webkit since its such a large project. Hopefully though (@kraj :-)) we can enable that to speed up rebuilds that include webkit, which would make rebuilds possible in <10 minutes.

**Comment by kraj** [comment]
my home desktop ( 44 cores) build WPE for me < 10 mins already :) on a serious note, I plan to look at it in coming week or so.

**Comment by hhromic** [comment]
thanks for the info, @wouterlucas !
I've been learning more about Yocto and I think I have a good idea for simplifying the customisation of the config.txt file.
I'm also going to enhance the overclocking mechanism, because the way is done currently overclocks always, no matter what machine you are building for.
For example in the case of the Rpi3, it actually lowers it down to 1Ghz.
I plan to put conditional variables to properly overclock to turbo profile depending on the target Rpi machine.
I already have good sources of suggestions for the best parameters for each machine type and I plan to incorporate that to be done automatically by the meta-wpe layer.
When I'm done with these ideas, I will send a pull-request for you to review and merge.
Cheers!

**Comment by hhromic** [comment]
Ah!
I saw you started working in a `wpe-eglfs-image` target to use EGLFS instead of Westeros.
I quickly tried to build an image with it out of curiosity and it didn't work because it tried to compile libprovision, which is private.
In case you are debugging! :)

**Comment by wouterlucas** [comment]
Just replace this line: https://github.com/WebPlatformForEmbedded/meta-wpe/blob/master/recipes-core/images/wpe-eglfs-image.bb#L21
with `packagegroup-wpe` it should build just fine (hope to share that soon!).

Add the following to your local.conf:
```
WPE_BACKEND = "rpi"
WPE_PLATFORM = "rpi"
DISTRO_FEATURES_remove = "wayland"
```

And enjoy an eglfs version without wayland :)

**Comment by hhromic** [comment]
Thanks @wouterlucas !
it's compiling like a charm with your input !

Guys, I just submitted a PR for fixing the `meta-raspberrypi` config file recipe as I described in this issue. After they merge it, I'm going to PR some changes I made to the `meta-wpe` side of the recipe to improve it.

I worked on the following changes:
* selectively enable Turbo overlocking depending on the target machine. Turbo must be configured differently if you build for RPI0, 1, RPI2 or RPI3. For example in RPI3 Turbo is not necessary and the way is done currently actually lower the RPI3 down. I'm basing the fine tuning on guides I've found.
* greatly simplified the rpi-config recipe and now all the default configuration is done in the `layer.conf` file, and the user can customise in `local.conf` (for example the `hdmi_*` options and splash screen enable/disable. I'm going to document this too.

Cheers and hope the `meta-raspberrypi` guys merge my PR soon to move on.
